// @huddle01/divergent-av/src/types/socket.types.ts
import { Server, Socket } from "socket.io";
import { SignalData } from "simple-peer";
import { IUserInfo } from "./common.types";
import { IErrorType } from "../store/storeTypes";

export interface ServerToClientEvents {
  "update-peer": (data: IUserInfo) => void;
  call: (data: { signal: SignalData; from: string; fromUid: string }) => void;
  answer: (signal: SignalData) => void;
  "end-call": (data: {}) => void;
  error: (data: { type: IErrorType; msg: string }) => void;
}

export interface ClientToServerEvents {
  "@update-peer": (data: {
    uid: string;
    id: string;
    type: string;
    avatar?: string;
    identityKey?: string;
    publicKey?: string;
    secretKey?: string;
  }) => void;
  "@call": (data: {
    peerToCall: string;
    signalData: SignalData;
    from: string;
  }) => void;
  "@answer": (data: { signal: SignalData; to?: string }) => void;
  "@end": (data: { peerToEnd: string | undefined }) => void;
}

export type IOServerType = Server<ClientToServerEvents, ServerToClientEvents>;

export type ISocketType = Socket<ClientToServerEvents, ServerToClientEvents>;
